File: matlab/mid/1d/obtain_1ring_curv_NM.m
function [ngbvs, nverts] = obtain_1ring_curv_NM...
    ( vid, edges, sibhvs, v2hv, varargin) %#codegen 
%OBTAIN_1RING_CURV Collect 1-ring vertices on non-manifold mesh.
% [NGBVS,NVERTS] = OBTAIN_1RING_CURV_NM(VID,EDGS,SIBHVS,V2HV) Collects 1-ring 
% vertices and edges of a vertex and saves them into NGBVS and NGBES. 
%
% See also OBTAIN_1RING_SURF, OBTAIN_1RING_VOL
%#codegen -args {int32(0), coder.typeof(int32(0), [inf,2]),coder.typeof(int32(0), [inf,2]),coder.typeof(int32(0), [inf,1])}
%#codegen obtain_1ring_curv_NM_usestruct -args
%#codegen {int32(0), coder.typeof(int32(0), [inf,2]),
%#codegen struct('eid',coder.typeof(int32(0), [inf,2]),  'lvid',coder.typeof(int8(0), [inf,2])),
%#codegen struct('eid',coder.typeof(int32(0), [inf,1]),  'lvid',coder.typeof(int8(0), [inf,1])),
%#codegen true}
coder.extrinsic('warning');
%assert( isa(vid, 'int32') && isa( edges, 'int32') && ...
%    isa( sibhvs, 'int32') && isa( v2hv,'int32'));

if nargin<5 || isempty(varargin{1}) || ~islogical(varargin{1})
    eid = hvid2eid(v2hv(vid));
    lid = hvid2lvid(v2hv(vid));
else
    eid = v2hv.eid(vid);
    lid = v2hv.lvid(vid);
end

MAXVALENCE=10;
ngbvs = zeros(MAXVALENCE,1,'int32');
nverts = int32(0);

if ~eid; return; end

% Collect one-ring vertices and edges
v = edges(eid, 3-lid); % another end of the edge
nverts = int32(1); ngbvs( nverts) = v;
if nargin<5 || isempty(varargin{1}) || ~islogical(varargin{1})
    opp = sibhvs(eid, lid);
else
    opp_eid=sibhvs.eid(eid, lid);
    opp_lvid=sibhvs.lvid(eid, lid);
end
if nargin<5 || isempty(varargin{1}) || ~islogical(varargin{1})
    while opp && opp~=v2hv(vid)
        if (nverts==MAXVALENCE); warning('MATLAB:MAXVALENCE','vertex %d valence exceeds MAXVALENCE=%d',vid,MAXVALENCE); end;
        
        eid = hvid2eid(opp); lid = hvid2lvid(opp);
        
        v = edges(eid, 3-lid); nverts = nverts+1; ngbvs( nverts) = v;
        
        % Next edge
        opp = sibhvs(eid, lid);
    end
else 
    while opp_eid && (opp_eid~=v2hv.eid(vid) || opp_lvid~=v2hv.lvid(vid))
        if (nverts==MAXVALENCE); warning('MATLAB:MAXVALENCE','vertex %d valence exceeds MAXVALENCE=%d',vid,MAXVALENCE); end;
        
        v = edges(opp_eid, 3-opp_lvid); nverts = nverts+1; ngbvs( nverts) = v;
       
        % Next edge
        opp_eid=sibhvs.eid(opp_eid, opp_lvid);
        opp_lvid=sibhvs.lvid(opp_eid, opp_lvid);
    end
end

File: matlab/mid/1d/codegen/lib/obtain_1ring_curv_NM/obtain_1ring_curv_NM.c
#include "rt_nonfinite.h"
#include "obtain_1ring_curv_NM.h"
#include "m2c.h"

void obtain_1ring_curv_NM(int32_T vid, const emxArray_int32_T *edges, const
  emxArray_int32_T *sibhvs, const emxArray_int32_T *v2hv, int32_T ngbvs[10],
  int32_T *nverts)
{
  int32_T eid;
  int32_T lid;
  int32_T i;
  eid = (int32_T)((uint32_T)v2hv->data[vid - 1] >> 1U);
  lid = (int32_T)((uint32_T)v2hv->data[vid - 1] & 1U);
  for (i = 0; i < 10; i++) {
    ngbvs[i] = 0;
  }

  *nverts = 0;
  if (!(eid != 0)) {
  } else {
    *nverts = 1;
    ngbvs[0] = edges->data[(eid + edges->size[0] * (1 - lid)) - 1];
    i = sibhvs->data[(eid + sibhvs->size[0] * lid) - 1];
    while ((i != 0) && (i != v2hv->data[vid - 1])) {
      eid = (int32_T)((uint32_T)i >> 1U) - 1;
      lid = (int32_T)((uint32_T)i & 1U);
      (*nverts)++;
      ngbvs[*nverts - 1] = edges->data[eid + edges->size[0] * (1 - lid)];
      i = sibhvs->data[eid + sibhvs->size[0] * lid];
    }
  }
}

void obtain_1ring_curv_NM_initialize(void)
{
  rt_InitInfAndNaN(8U);
}

void obtain_1ring_curv_NM_terminate(void)
{
}

void obtain_1ring_curv_NM_usestruct(int32_T vid, const emxArray_int32_T *edges,
  const struct_T sibhvs, const struct_T v2hv, boolean_T usestruct, int32_T
  ngbvs[10], int32_T *nverts)
{
  int32_T i;
  int8_T opp_lvid;
  for (i = 0; i < 10; i++) {
    ngbvs[i] = 0;
  }

  *nverts = 0;
  if (!(v2hv.eid->data[vid - 1] != 0)) {
  } else {
    *nverts = 1;
    ngbvs[0] = edges->data[(v2hv.eid->data[vid - 1] + edges->size[0] * ((int8_T)
                             (3 - v2hv.lvid->data[vid - 1]) - 1)) - 1];
    i = sibhvs.eid->data[(v2hv.eid->data[vid - 1] + sibhvs.eid->size[0] *
                          (v2hv.lvid->data[vid - 1] - 1)) - 1];
    opp_lvid = sibhvs.lvid->data[(v2hv.eid->data[vid - 1] + sibhvs.lvid->size[0]
      * (v2hv.lvid->data[vid - 1] - 1)) - 1];
    while ((i != 0) && ((i != v2hv.eid->data[vid - 1]) || (opp_lvid !=
             v2hv.lvid->data[vid - 1]))) {
      (*nverts)++;
      ngbvs[*nverts - 1] = edges->data[(i + edges->size[0] * ((int8_T)(3 -
        opp_lvid) - 1)) - 1];
      i = sibhvs.eid->data[(i + sibhvs.eid->size[0] * (opp_lvid - 1)) - 1];
      opp_lvid = sibhvs.lvid->data[(i + sibhvs.lvid->size[0] * (opp_lvid - 1)) -
        1];
    }
  }
}
